### PR TITLE
[NFR] Add the ability to use FrontendInterface to serialize Model and ResultSet

### PR DESCRIPTION
**Pull request by ogarbe**
Hello!

* Type: new feature
* Link to issue: #11999 

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)?
- [x] I have checked that another pull request for this purpose does not exist.
- [ ] I wrote some tests for this PR.

Small description of change: add FrontendInterface in DI (name "serializer") in order to serialize/unserialize Model, Collection, Resultset with a better thing than serialize/unserialize method

Thanks

### CODE REVIEW

**Comment by sergeyklay** [comment]
We don't add changes directly to stable branches or `master`. Could you please send this PR to the `3.2.x` branch?

File: phalcon/mvc/model/resultset/complex.zep [hunk 1]
```diff
@@ -282,30 +283,42 @@ class Complex extends Resultset implements ResultsetInterface
 		let cache = this->_cache,
 			columnTypes = this->_columnTypes,
 			hydrateMode = this->_hydrateMode;
+		let dependencyInjector = \Phalcon\DI::getDefault();
```
Comment (sergeyklay) [inline phalcon/mvc/model/resultset/complex.zep]:
```zep
use Phalcon\Di;

// ...

let dependencyInjector = Di::getDefault();

if typeof dependencyInjector != "object" {

}
```

File: phalcon/mvc/model/resultset/complex.zep [hunk 2]
```diff
@@ -282,30 +283,42 @@ class Complex extends Resultset implements ResultsetInterface
...
-		var resultset;
-
+		var resultset, dependencyInjector, serializer;
 		/**
 		* Rows are already hydrated
 		*/
 		let this->_disableHydration = true;
-
-		let resultset = unserialize(data);
+		let dependencyInjector = \Phalcon\DI::getDefault();
```
Comment (sergeyklay) [inline phalcon/mvc/model/resultset/complex.zep]: The same here

File: phalcon/mvc/model/resultset/simple.zep [hunk 3]
```diff
@@ -245,9 +261,14 @@ class Simple extends Resultset
 	 */
 	public function unserialize(string! data) -> void
 	{
-		var resultset, keepSnapshots;
-
-		let resultset = unserialize(data);
+		var resultset, keepSnapshots, dependencyInjector, serializer;
+		let dependencyInjector = \Phalcon\DI::getDefault();
```
Comment (sergeyklay) [inline phalcon/mvc/model/resultset/simple.zep]: The same here

File: phalcon/mvc/model/resultset/simple.zep [hunk 4]
```diff
@@ -227,6 +230,19 @@ class Simple extends Resultset
 	 */
 	public function serialize() -> string
 	{
+		var dependencyInjector, serializer;
+		let dependencyInjector = \Phalcon\DI::getDefault();
```
Comment (sergeyklay) [inline phalcon/mvc/model/resultset/simple.zep]: the same here

**Review by sergeyklay — CHANGES_REQUESTED** [review]

**Comment by ogarbe** [comment]
@sergeyklay CR done + PR changed

**Review by sergeyklay — DISMISSED** [review]

**Comment by rrelenschi** [comment]
Hello,

@sergeyklay , @ogarbe 
I've added a small change, to check that serializer is an object before using "instanceof" because calling instanceof on non-objects doesn't work.
Please review : https://github.com/phalcon/cphalcon/pull/12808/commits/c31caf01fbe39fc8df2108babd2208ee8ad36007
Thanks !

**Review by ogarbe — COMMENTED** [review]
ok for me

(i can't approve)

**Comment by ogarbe** [comment]
Hello, why hasn't this been merged into 3.2 ? This feature increases a lot performances and has complete fallback !

**Comment by sergeyklay** [comment]
@ogarbe Actually I can't merge this because this PR introduces segfaults. I'll tale a look later

**Comment by ogarbe** [comment]
@sergeyklay any update please ?

**Comment by sergeyklay** [comment]
@ogarbe Could you please separate this PR to few small PRs to detect the root of problem. Note: you have to use the 4.x branch

**Comment by virgofx** [comment]
Model is already serializable... shouldn't need to do anything..

**Comment by ogarbe** [comment]
@sergeyklay it was a small PR...

@virgofx yes it is, but it uses the PHP serializer and not the serializer we can define everywhere else in Phalcon... Hence poor performance (ig_serializer is 3 times faster...)

**Comment by ogarbe** [comment]
@sergeyklay any help i can do in order to merge this for next release ?

**Comment by sergeyklay** [comment]
@ogarbe It would be helpful to separate this PR into several small

**Comment by ogarbe** [comment]
@sergeyklay i don't understand : the PR just 
- add a serializer in the DI
- use the serializer in model and resultset
- use the serializer in collection

What split would you like ? Adding the serializer without using it is a nonsense. Adding the serializer and using it for model but no resultset is also a nonsense. So i could split to remove collection, but i don't get what would be the use...

**Comment by sergeyklay** [comment]
@ogarbe Do you know which change produce segfault? I mean file and line number. How are you going to find the issue?

**Comment by ogarbe** [comment]
i'm not aware of any segfault, can you give me info about it, you told me this

``
@ogarbe Actually I can't merge this because this PR introduces segfaults. I'll take a look a bit later
--

and no more info. If you could tell me more, i would be able to.

**Comment by sergeyklay** [comment]
@ogarbe 
As you can see this PR marked for the `4.0.x` branch. There is no reason to resolve this ASAP. The `3.2.x` branch only in the security support stage now.  So we don't plan to release `4.0.0` tomorrow or a mont or even in a month. Let me deal with this a bit later. In any case, thanks for your contribution and for the fact that you are so patiently ping me.

**Comment by ogarbe** [comment]
@sergeyklay **9 month** after and this PR has not been looked at. It was ok on 3.1 than 3.2...

You rolled-out 3.3 which have new features (case / switch on volt) but not this one which increases performance on ALL storage on MVC requests.

What's the problem ?.

**Comment by sergeyklay** [comment]
@ogarbe Thanks for the reminder. I apologize for completely overlooking this. Thank you for the PR. I know some people think this kinds of changes are trivial, but personally I believe it is important to get everything right, including tests and general design. There were some problems here and for a long time I couldn't make it work as expected. At least some tests constantly was failed. We can't accept new features in minor versions. Next major version is 4.0.0. Could you send your PR to the 4.0.x branch. So thanks again for this New Feature Request, and again I am sorry about the delay.

**Comment by ogarbe** [comment]
@sergeyklay : PR updated to 4.0.0

**Comment by niden** [comment]
@ogarbe Thank you for this. Sorry to be a pain, can you resolve the conflict please? Assigning this to me also so that I can check it out with @sergeyklay when he is available.

**Comment by sergeyklay** [comment]
I canceled AppVeyor's build for this PR. No worries, I just need a spare build agent right now. I'll restart this build a bit later

**Comment by ogarbe** [comment]
> @ogarbe Thank you for this. Sorry to be a pain, can you resolve the conflict please? Assigning this to me also so that I can check it out with @sergeyklay when he is available.

it's done @niden

**Comment by niden** [comment]
@ogarbe One last thing please. Can you add a line to the `CHANGELOG-4` please regarding this feature?

This is solid.

**Comment by ogarbe** [comment]
@niden : done

**Review by niden — APPROVED** [review]

**Pull request merged**